File: sdks/php/src/Retry.php
<?php

declare(strict_types=1);

namespace Nexus\SDK;

use GuzzleHttp\Exception\ConnectException;
use GuzzleHttp\Exception\RequestException;

/**
 * Configuration for retry behavior.
 */
class RetryConfig
{
    /**
     * @param int[] $retryableStatusCodes HTTP status codes that should trigger a retry
     */
    public function __construct(
        /** Maximum number of retry attempts (default: 3) */
        public int $maxRetries = 3,
        /** Initial backoff duration in milliseconds (default: 100) */
        public int $initialBackoffMs = 100,
        /** Maximum backoff duration in milliseconds (default: 10000) */
        public int $maxBackoffMs = 10000,
        /** Multiplier for exponential backoff (default: 2.0) */
        public float $backoffMultiplier = 2.0,
        /** Whether to add jitter to backoff (default: true) */
        public bool $jitter = true,
        /** HTTP status codes that should trigger a retry */
        public array $retryableStatusCodes = [
            408, // Request Timeout
            429, // Too Many Requests
            500, // Internal Server Error
            502, // Bad Gateway
            503, // Service Unavailable
            504, // Gateway Timeout
        ]
    ) {
    }

    /**
     * Create a default retry configuration.
     */
    public static function default(): self
    {
        return new self();
    }

    /**
     * Check if an exception is retryable.
     */
    public function isRetryable(\Throwable $e): bool
    {
        // Connection errors are retryable
        if ($e instanceof ConnectException) {
            return true;
        }

        // Check HTTP status codes
        if ($e instanceof NexusApiException) {
            return in_array($e->statusCode, $this->retryableStatusCodes, true);
        }

        if ($e instanceof RequestException && $e->hasResponse()) {
            $statusCode = $e->getResponse()?->getStatusCode();
            return $statusCode !== null && in_array($statusCode, $this->retryableStatusCodes, true);
        }

        return false;
    }

    /**
     * Calculate backoff duration for a given attempt.
     */
    public function calculateBackoffMs(int $attempt): int
    {
        $backoff = $this->initialBackoffMs * pow($this->backoffMultiplier, $attempt);

        if ($this->jitter) {
            // Add ±25% jitter
            $jitterRange = $backoff * 0.25;
            $backoff = $backoff - $jitterRange + (mt_rand() / mt_getrandmax() * $jitterRange * 2);
        }

        return min((int) $backoff, $this->maxBackoffMs);
    }
}

/**
 * A wrapper around NexusClient that adds automatic retry functionality.
 */
class RetryableClient
{
    private NexusClient $client;
    private RetryConfig $retryConfig;

    public function __construct(Config $config, ?RetryConfig $retryConfig = null)
    {
        $this->client = new NexusClient($config);
        $this->retryConfig = $retryConfig ?? RetryConfig::default();
    }

    /**
     * Create from an existing client.
     */
    public static function fromClient(NexusClient $client, ?RetryConfig $retryConfig = null): self
    {
        $instance = new self(new Config(), $retryConfig);
        $instance->client = $client;
        return $instance;
    }

    /**
     * Set bearer token for authentication.
     */
    public function setToken(string $token): void
    {
        $this->client->setToken($token);
    }

    /**
     * Execute an operation with automatic retry.
     *
     * @template T
     * @param callable(): T $operation
     * @return T
     * @throws \Throwable
     */
    private function executeWithRetry(callable $operation): mixed
    {
        $lastException = null;

        for ($attempt = 0; $attempt <= $this->retryConfig->maxRetries; $attempt++) {
            try {
                return $operation();
            } catch (\Throwable $e) {
                $lastException = $e;

                if (!$this->retryConfig->isRetryable($e)) {
                    throw $e;
                }

                if ($attempt < $this->retryConfig->maxRetries) {
                    $backoffMs = $this->retryConfig->calculateBackoffMs($attempt);
                    usleep($backoffMs * 1000); // Convert to microseconds
                }
            }
        }

        throw $lastException ?? new NexusException('Retry failed without exception');
    }

    /**
     * Check if server is reachable with automatic retry.
     *
     * @throws NexusApiException
     */
    public function ping(): void
    {
        $this->executeWithRetry(fn() => $this->client->ping());
    }

    /**
     * Execute a Cypher query with automatic retry.
     *
     * @param array<string, mixed>|null $parameters
     * @throws NexusApiException
     */
    public function executeCypher(string $query, ?array $parameters = null): QueryResult
    {
        return $this->executeWithRetry(fn() => $this->client->executeCypher($query, $parameters));
    }

    /**
     * Create a new node with automatic retry.
     *
     * @param string[] $labels
     * @param array<string, mixed> $properties
     * @throws NexusApiException
     */
    public function createNode(array $labels, array $properties): Node
    {
        return $this->executeWithRetry(fn() => $this->client->createNode($labels, $properties));
    }

    /**
     * Get a node by ID with automatic retry.
     *
     * @throws NexusApiException
     */
    public function getNode(string $id): Node
    {
        return $this->executeWithRetry(fn() => $this->client->getNode($id));
    }

    /**
     * Update a node's properties with automatic retry.
     *
     * @param array<string, mixed> $properties
     * @throws NexusApiException
     */
    public function updateNode(string $id, array $properties): Node
    {
        return $this->executeWithRetry(fn() => $this->client->updateNode($id, $properties));
    }

    /**
     * Delete a node with automatic retry.
     *
     * @throws NexusApiException
     */
    public function deleteNode(string $id): void
    {
        $this->executeWithRetry(fn() => $this->client->deleteNode($id));
    }

    /**
     * Create a relationship with automatic retry.
     *
     * @param array<string, mixed> $properties
     * @throws NexusApiException
     */
    public function createRelationship(
        string $startNode,
        string $endNode,
        string $type,
        array $properties
    ): Relationship {
        return $this->executeWithRetry(
            fn() => $this->client->createRelationship($startNode, $endNode, $type, $properties)
        );
    }

    /**
     * List all labels with automatic retry.
     *
     * @return string[]
     * @throws NexusApiException
     */
    public function listLabels(): array
    {
        return $this->executeWithRetry(fn() => $this->client->listLabels());
    }

    /**
     * List all relationship types with automatic retry.
     *
     * @return string[]
     * @throws NexusApiException
     */
    public function listRelationshipTypes(): array
    {
        return $this->executeWithRetry(fn() => $this->client->listRelationshipTypes());
    }

    /**
     * List all indexes with automatic retry.
     *
     * @return Index[]
     * @throws NexusApiException
     */
    public function listIndexes(): array
    {
        return $this->executeWithRetry(fn() => $this->client->listIndexes());
    }

    /**
     * Get the underlying client.
     */
    public function getClient(): NexusClient
    {
        return $this->client;
    }
}

File: sdks/csharp/Retry.cs
using System.Net;

namespace Nexus.SDK;

/// <summary>
/// Configuration for retry behavior.
/// </summary>
public class RetryConfig
{
    /// <summary>
    /// Maximum number of retry attempts (default: 3).
    /// </summary>
    public int MaxRetries { get; set; } = 3;

    /// <summary>
    /// Initial backoff duration (default: 100ms).
    /// </summary>
    public TimeSpan InitialBackoff { get; set; } = TimeSpan.FromMilliseconds(100);

    /// <summary>
    /// Maximum backoff duration (default: 10s).
    /// </summary>
    public TimeSpan MaxBackoff { get; set; } = TimeSpan.FromSeconds(10);

    /// <summary>
    /// Multiplier for exponential backoff (default: 2.0).
    /// </summary>
    public double BackoffMultiplier { get; set; } = 2.0;

    /// <summary>
    /// Whether to add jitter to backoff (default: true).
    /// </summary>
    public bool Jitter { get; set; } = true;

    /// <summary>
    /// HTTP status codes that should trigger a retry.
    /// </summary>
    public HashSet<HttpStatusCode> RetryableStatusCodes { get; set; } = new()
    {
        HttpStatusCode.RequestTimeout,        // 408
        HttpStatusCode.TooManyRequests,       // 429
        HttpStatusCode.InternalServerError,   // 500
        HttpStatusCode.BadGateway,            // 502
        HttpStatusCode.ServiceUnavailable,    // 503
        HttpStatusCode.GatewayTimeout         // 504
    };

    /// <summary>
    /// Creates a default retry configuration.
    /// </summary>
    public static RetryConfig Default => new();
}

/// <summary>
/// Extension methods for adding retry support to NexusClient.
/// </summary>
public static class RetryExtensions
{
    private static readonly Random Random = new();

    /// <summary>
    /// Creates a retryable client wrapper.
    /// </summary>
    public static RetryableNexusClient WithRetry(this NexusClient client, RetryConfig? config = null)
    {
        return new RetryableNexusClient(client, config ?? RetryConfig.Default);
    }

    internal static bool IsRetryableException(this RetryConfig config, Exception ex)
    {
        if (ex is NexusApiException apiEx)
        {
            return config.RetryableStatusCodes.Contains((HttpStatusCode)apiEx.StatusCode);
        }

        // Network errors and timeouts are retryable
        return ex is HttpRequestException or TaskCanceledException or OperationCanceledException;
    }

    internal static TimeSpan CalculateBackoff(this RetryConfig config, int attempt)
    {
        var backoff = config.InitialBackoff.TotalMilliseconds * Math.Pow(config.BackoffMultiplier, attempt);

        if (config.Jitter)
        {
            // Add ±25% jitter
            var jitterRange = backoff * 0.25;
            backoff = backoff - jitterRange + (Random.NextDouble() * jitterRange * 2);
        }

        var duration = TimeSpan.FromMilliseconds(backoff);
        return duration > config.MaxBackoff ? config.MaxBackoff : duration;
    }
}

/// <summary>
/// A wrapper around NexusClient that adds automatic retry functionality.
/// </summary>
public class RetryableNexusClient : IDisposable
{
    private readonly NexusClient _client;
    private readonly RetryConfig _retryConfig;
    private bool _disposed;

    /// <summary>
    /// Creates a new retryable client.
    /// </summary>
    public RetryableNexusClient(NexusClient client, RetryConfig? config = null)
    {
        _client = client ?? throw new ArgumentNullException(nameof(client));
        _retryConfig = config ?? RetryConfig.Default;
    }

    /// <summary>
    /// Creates a new retryable client with configuration.
    /// </summary>
    public RetryableNexusClient(NexusClientConfig clientConfig, RetryConfig? retryConfig = null)
    {
        _client = new NexusClient(clientConfig);
        _retryConfig = retryConfig ?? RetryConfig.Default;
    }

    /// <summary>
    /// Executes an operation with automatic retry.
    /// </summary>
    private async Task<T> ExecuteWithRetryAsync<T>(
        Func<CancellationToken, Task<T>> operation,
        CancellationToken cancellationToken = default)
    {
        Exception? lastException = null;

        for (var attempt = 0; attempt <= _retryConfig.MaxRetries; attempt++)
        {
            cancellationToken.ThrowIfCancellationRequested();

            try
            {
                return await operation(cancellationToken);
            }
            catch (Exception ex) when (_retryConfig.IsRetryableException(ex))
            {
                lastException = ex;

                if (attempt < _retryConfig.MaxRetries)
                {
                    var backoff = _retryConfig.CalculateBackoff(attempt);
                    await Task.Delay(backoff, cancellationToken);
                }
            }
        }

        throw lastException ?? new InvalidOperationException("Retry failed without exception");
    }

    /// <summary>
    /// Executes an operation with automatic retry (void return).
    /// </summary>
    private async Task ExecuteWithRetryAsync(
        Func<CancellationToken, Task> operation,
        CancellationToken cancellationToken = default)
    {
        await ExecuteWithRetryAsync(async ct =>
        {
            await operation(ct);
            return true;
        }, cancellationToken);
    }

    #region Health Check

    /// <summary>
    /// Checks if the server is reachable with automatic retry.
    /// </summary>
    public async Task PingAsync(CancellationToken cancellationToken = default)
    {
        await ExecuteWithRetryAsync(
            ct => _client.PingAsync(ct),
            cancellationToken);
    }

    #endregion

    #region Cypher Queries

    /// <summary>
    /// Executes a Cypher query with automatic retry.
    /// </summary>
    public async Task<QueryResult> ExecuteCypherAsync(
        string query,
        Dictionary<string, object?>? parameters = null,
        CancellationToken cancellationToken = default)
    {
        return await ExecuteWithRetryAsync(
            ct => _client.ExecuteCypherAsync(query, parameters, ct),
            cancellationToken);
    }

    #endregion

    #region Node Operations

    /// <summary>
    /// Creates a new node with automatic retry.
    /// </summary>
    public async Task<Node> CreateNodeAsync(
        List<string> labels,
        Dictionary<string, object?> properties,
        CancellationToken cancellationToken = default)
    {
        return await ExecuteWithRetryAsync(
            ct => _client.CreateNodeAsync(labels, properties, ct),
            cancellationToken);
    }

    /// <summary>
    /// Retrieves a node by ID with automatic retry.
    /// </summary>
    public async Task<Node> GetNodeAsync(
        string id,
        CancellationToken cancellationToken = default)
    {
        return await ExecuteWithRetryAsync(
            ct => _client.GetNodeAsync(id, ct),
            cancellationToken);
    }

    /// <summary>
    /// Updates a node with automatic retry.
    /// </summary>
    public async Task<Node> UpdateNodeAsync(
        string id,
        Dictionary<string, object?> properties,
        CancellationToken cancellationToken = default)
    {
        return await ExecuteWithRetryAsync(
            ct => _client.UpdateNodeAsync(id, properties, ct),
            cancellationToken);
    }

    /// <summary>
    /// Deletes a node with automatic retry.
    /// </summary>
    public async Task DeleteNodeAsync(
        string id,
        CancellationToken cancellationToken = default)
    {
        await ExecuteWithRetryAsync(
            ct => _client.DeleteNodeAsync(id, ct),
            cancellationToken);
    }

    #endregion

    #region Schema Management

    /// <summary>
    /// Retrieves all node labels with automatic retry.
    /// </summary>
    public async Task<List<string>> ListLabelsAsync(
        CancellationToken cancellationToken = default)
    {
        return await ExecuteWithRetryAsync(
            ct => _client.ListLabelsAsync(ct),
            cancellationToken);
    }

    /// <summary>
    /// Retrieves all relationship types with automatic retry.
    /// </summary>
    public async Task<List<string>> ListRelationshipTypesAsync(
        CancellationToken cancellationToken = default)
    {
        return await ExecuteWithRetryAsync(
            ct => _client.ListRelationshipTypesAsync(ct),
            cancellationToken);
    }

    /// <summary>
    /// Retrieves all indexes with automatic retry.
    /// </summary>
    public async Task<List<Index>> ListIndexesAsync(
        CancellationToken cancellationToken = default)
    {
        return await ExecuteWithRetryAsync(
            ct => _client.ListIndexesAsync(ct),
            cancellationToken);
    }

    #endregion

    #region IDisposable

    protected virtual void Dispose(bool disposing)
    {
        if (!_disposed)
        {
            if (disposing)
            {
                _client.Dispose();
            }
            _disposed = true;
        }
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    #endregion
}
